forest/benchmarking/observable_estimation.py

```diff
@@ -900,7 +901,15 @@ def estimate_observables(qc: QuantumComputer, obs_expt: ObservablesExperiment,
     programs, meas_qubits = generate_experiment_programs(obs_expt, active_reset)
     for prog, meas_qs, settings in zip(tqdm(programs, disable=not show_progress_bar), meas_qubits,
                                        obs_expt):
-        results = qc.run_symmetrized_readout(prog, num_shots, symm_type, meas_qs)
+        # Remove `Delay` instructions for qvm-bound executions,
+        # where Quil-T instructions are not supported.
+        if 'qvm' in qc.name:
```

Comment:
I would strongly prefer that pyquil be responsible to ensuring we send QVM-compatible quil to the QVM.

Reply:
In the interest of keeping behavior explicit, do you mean that there should be something like one of the following?

```py
# Configure a preprocessing layer
qc = get_qc("9q-qvm", automatically_adapt_programs_for_compatibiliity=True)

# Explicitly adapt, implicitly transform
program = program.adapt_for_qvm()

# Explicitly transform, implicitly adapt
program = program.remove_quilt()
```

Reply:
I may be wrong about this, but I believe the QVM supports "protoquil". So maybe something like:

```python
program = program.to_protoquil()
```

My instinct would be to put it right here: https://github.com/rigetti/pyquil/blob/a72499c8a0e44ed09346cef3e33234bac7ca829d/pyquil/api/_qvm.py#L150

There is already a choice about how to express the program with `executable.out(calibrations=False)`. As best I can tell, it's simply an insufficient one. Is there any reason we should not remove all the QVM-incompatible instructions before submission?

Reply:
I'm not familiar enough with the semantics of the incompatible instructions to assess the risk that a user's program might be significantly changed by automatic removal of them, versus the user's burden of correcting errors returned by the QVM. If that risk is nonexistent like ignoring `calibrations` as I think you're suggesting then I agree with you, otherwise I'd worry about confusion. There's also the (probably quite small) chance of later QVM support falling out of sync with pyquil's understanding of it

Reply:
`executable.out(calibrations=False)` removes `DEFCAL`s, which, being global state, can be removed before passing the program to `quilc` and then easily restored after. This is in fact one of the ways pyQuil uses the flag. The problem with doing this for all Quil-T instructions is that where they appear in the program is important, and it's not possible to reliably put them in the right place after compiling with `quilc`. Of course, we could fix this issue for QVM runs specifically, but that would be a bit fragile, and I think adding these hidden behaviors doubles down on the wrong pattern. They make pyQuil more complex from a maintenance and user perspective. It's a lot harder to figure out what the rules are if the system tries to work around all of them for you. That is fine, up until one of the rules is important to you, then you have to unravel the system to try and figure out where it's making a bad decision for you. In general, I think we should be more explicit about the limitations of our stack so that users become knowledgeable about where those limitations and rough edges are and instead provide the tools needed to work with them. At the end of the day, running programs on a QVM is different from a QPU, and I'd imagine they are likely to become _more_ different as time goes on. We can continue to push these complexities into pyQuil, but I think it's more empowering to give users the information and tools to handle these cases, rather than making them entirely dependent on us to sweep them under the rug. 

As an idea, we can already discern whether a QC is targeting a QVM or QPU by looking at its compiler. We can use that to make an informed decision about whether or not to remove Quil-T instructions from the program. This is already possible with an `ininstance` check, but we could make this even easier by adding some helper that makes it easier to tell if some `AbstractInstruction` is a Quil-T instruction.

```python
program = ...
if isinstance(qc.compiler, QVMCompiler):
    # Targeting a QVM, which doesn't 
    program = [inst for inst in program if not isinstance(inst, (Delay, Fence, ...))]
```

Reply:
I've opened https://github.com/rigetti/pyquil/issues/1684 for further discussion, since @MarquessV I'm not sure if your suggestion involves `pyquil` or dependent library changes

Reply:
Now that https://github.com/rigetti/pyquil/pull/1718 is resolved, I'll update the pyquil dependency and use that.